File: Algorithms/1306_Jump_Game_III/Python/Jump_Game_III_Solution_1.py

# Space: O(n)
# Time: O(n)

# BFS Approach

class Solution:
    def canReach(self, arr, start):
        length = len(arr)
        if length == 1 and start == 0: return True
        visited = set()
        queue = [start]

        while queue:
            cur = queue.pop(0)
            if cur in visited: continue
            if queue[cur] == 0: return True
            visited.add(cur)
            if 0 <= cur + queue[cur] < length: queue.append(cur + queue[cur])
            if 0 <= cur - queue[cur] < length: queue.append(cur - queue[cur])

        return False

File: Algorithms/1306_Jump_Game_III/Java/JumpGameIIISolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public boolean canReach(int[] arr, int start) {
        if (arr.length == 1) {
            return arr[0] == 0;
        }
        HashSet<Integer> cache = new HashSet<>();
        cache.add(start);
        Queue<Integer> queue = new ArrayDeque<>();
        queue.add(start);
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            if (arr[cur] == 0) return true;
            int left = cur - arr[cur];
            int right = cur + arr[cur];

            if (0 <= left && left < arr.length && !cache.contains(left)) {
                queue.add(left);
                cache.add(left);
            }
            if (0 <= right && right < arr.length && !cache.contains(right)) {
                queue.add(right);
                cache.add(right);
            }
        }
        return false;
    }
}




